exchanges/okgroup/okgroup_websocket.go

```diff
@@ -292,113 +233,174 @@ func (o *OKGroup) WsLogin() error {
 	return nil
 }
 
-// WsHandleErrorResponse sends an error message to ws handler
-func (o *OKGroup) WsHandleErrorResponse(event WebsocketErrorResponse) {
-	errorMessage := fmt.Sprintf("%v error - %v message: %s ",
-		o.Name,
-		event.ErrorCode,
-		event.Message)
-	if o.Verbose {
-		log.Error(log.ExchangeSys, errorMessage)
+// wsReadData receives and passes on websocket messages for processing
+func (o *OKGroup) WsReadData(wg *sync.WaitGroup) {
+	o.Websocket.Wg.Add(1)
+	defer func() {
+		o.Websocket.Wg.Done()
+	}()
+	wg.Done()
+
+	for {
+		select {
+		case <-o.Websocket.ShutdownC:
+			return
+		default:
+			resp, err := o.WebsocketConn.ReadMessage()
+			if err != nil {
+				o.Websocket.ReadMessageErrors <- err
+				return
+			}
+			o.Websocket.TrafficAlert <- struct{}{}
+			err = o.WsHandleData(resp.Raw)
+			if err != nil {
+				o.Websocket.DataHandler <- err
+			}
+		}
 	}
-	o.Websocket.DataHandler <- fmt.Errorf(errorMessage)
 }
 
-// GetWsChannelWithoutOrderType takes WebsocketDataResponse.Table and returns
-// The base channel name eg receive "spot/depth5:BTC-USDT" return "depth5"
-func (o *OKGroup) GetWsChannelWithoutOrderType(table string) string {
-	index := strings.Index(table, "/")
-	if index == -1 {
-		return table
-	}
-	channel := table[index+1:]
-	index = strings.Index(channel, ":")
-	// Some events do not contain a currency
-	if index == -1 {
-		return channel
+// WsHandleData will read websocket raw data and pass to appropriate handler
+func (o *OKGroup) WsHandleData(respRaw []byte) error {
+	var dataResponse WebsocketDataResponse
+	err := json.Unmarshal(respRaw, &dataResponse)
+	if err != nil {
+		return err
 	}
+	if len(dataResponse.Data) > 0 {
+		switch o.GetWsChannelWithoutOrderType(dataResponse.Table) {
+		case okGroupWsCandle60s, okGroupWsCandle180s, okGroupWsCandle300s,
+			okGroupWsCandle900s, okGroupWsCandle1800s, okGroupWsCandle3600s,
+			okGroupWsCandle7200s, okGroupWsCandle14400s, okGroupWsCandle21600s,
+			okGroupWsCandle43200s, okGroupWsCandle86400s, okGroupWsCandle604900s:
+			return o.wsProcessCandles(respRaw)
+		case okGroupWsDepth, okGroupWsDepth5:
+			return o.WsProcessOrderBook(respRaw)
+		case okGroupWsTicker:
+			return o.wsProcessTickers(respRaw)
+		case okGroupWsTrade:
+			return o.wsProcessTrades(respRaw)
+		case okGroupWsOrder:
+			return o.wsProcessOrder(respRaw)
+		}
+		o.Websocket.DataHandler <- wshandler.UnhandledMessageWarning{Message: o.Name + wshandler.UnhandledMessage + string(respRaw)}
+		return nil
+	}
+
+	var errorResponse WebsocketErrorResponse
+	err = json.Unmarshal(respRaw, &errorResponse)
+	if err == nil && errorResponse.ErrorCode > 0 {
+		if o.Verbose {
+			log.Debugf(log.ExchangeSys,
+				"WS Error Event: %v Message: %v for %s",
```

Comment:
To keep things consistent, please place the exchange name first like the following fmt.Errorf